### ISSUE TITLE
Rule idea: `dollar-variable-place`

### ISSUE THREAD

**Issue by evilebottnawi** [kristerkari/stylelint-scss#103]
Allow and disallow inside block or out side
- `dollar-variable-place: "outside-declaration-block"`

Valid:
```scss
$color: red;

a {
  color: $color;
}
```
Invalid:
```scss
a {
  $color: red;
  color: $color;
}
```
- `dollar-variable-place: "inside-declaration-block"`

Invalid:
```scss
$color: red;

a {
  color: $color;
}
```
Valid:
```scss
a {
  $color: red;
  color: $color;
}
```
Maybe best name for rule and options. I can do PR.

**Comment by jeddy3** [comment]
Thanks for the suggestion. I think this would make a good plugin candidate.

Perhaps within the `stylelint-scss` plugin pack. Quoting from their `README.md`:

> stylelint-scss introduces rules specific to SCSS syntax. That said, the rules from this plugin can be used with other syntaxes, like Less or some PostCSS syntaxes. That's why the rules' names are not tied to SCSS only (`at-function-pattern` instead of `scss-function-pattern`).

**Comment by jeddy3** [comment]
> Maybe best name for rule and options. I can do PR.

I think what you're trying to do is _disallow_ something. The stylelint [convention for that](https://github.com/stylelint/stylelint/blob/master/docs/user-guide/about-rules.md#no-rules) is to use a `no` rule. Perhaps `no-dollar-variable` would work here, along with two `ignore` options: `nested` and `non-nested`.

e.g. `no-dollar-variable: [true, { "ignore: ["nested"] }]`.

This would be consistent with any future rules that disallow scss-like constructs like `maps`, `lists` etc.

**Comment by evilebottnawi** [comment]
@jeddy3 sgtm to me, let's start `no-dollar-variable: [true, { "ignore: ["inside-block"] }]`?

**Comment by jeddy3** [comment]
`inside-block` is how we consistently label it in stylelint core, but if "global"/"local" is idomatic to scss then they're likely more fitting names for in this plugin pack.

**Comment by evilebottnawi** [comment]
@dryoma @jeddy3 final decision?

**Comment by jeddy3** [comment]
It's up to @dryoma. Just thought I'd chime in with how stylelint does it in core in case the info was helpful :)

**Comment by dryoma** [comment]
@kristerkari what do you think?

"global"/"local"
+ more semantic
- two rules instead of one (is it a con?)

"inside-block"
+ allows to go with just one rule
+ consistent with future rules for other constructs (if any)

**Comment by evilebottnawi** [comment]
@dryoma also be good allow some variable, example be good allow `$parent: &;` (`ignoreVariables: []`). It is normal using inside declaration block.

**Comment by kristerkari** [comment]
@dryoma oh, that's a tough one. I think that `global/local` makes more sense, but then we'd have 2 rules instead of one (does that make things more complicated for the user or not?).

**Comment by kristerkari** [comment]
@dryoma but also I think that there is a benefit in following the naming conventions from Stylelint. It might be easier for the users not having to switch between different naming conventions when using Stylelint plugins.

**Comment by dryoma** [comment]
> It might be easier for the users not having to switch between different naming conventions when using Stylelint plugins.

Thought so too, but not sure. And the question is how many such entities (to allow/disallow in a scopre) we'd need? I mean, all these extends and mixin calls/declarations are probably better off in stylelint as a rule for at-rules + specific/ignored names. What remains for us are $-variables, %-placeholders maybe. & doesn't make sense globally.

> we'd have 2 rules instead of one

And then I just had a thought: _do_ we need the 2 rules? Won't `no-dollar-variable`: `local/global` do the trick?

**Comment by kristerkari** [comment]
> And then I just had a thought: do we need the 2 rules? Won't no-dollar-variable: local/global do the trick?

True, I think that this is the best of the options we have.

**Comment by jeddy3** [comment]
`local/global` does make sense as that is the terminology used in the [SASS reference](http://sass-lang.com/documentation/file.SASS_REFERENCE.html#variables_).

> Won't no-dollar-variable: local/global do the trick?

Just to reiterate, idiomatically in stylelint `no` rules are most restrictive _by default_ and can be _made more permissive using `"ignore": []`_ options. 

So, I'd recommend _a single_ `no-dollar-variable` rule with `true` as the primary option and providing _two `ignore` keyword_ optional secondary options: `"global"` and `"local"`.

Disallow all dollar-variables:

```json
"no-dollar-variable": true
```

Disallow global dollar-variables:

```json
"no-dollar-variable": [true, { "ignore": ["local"] }]
```

Disallow local dollar-variables:

```json
"no-dollar-variable": [true, { "ignore": ["global"] }]
```

This is how we'd idiomatically do it in stylelint.

As an aside, and you probably realise this already but I just noticed in the SASS reference, it looks like you'll need to account for the `!global` flag when determining if a dollar-variable is local or global.

**Comment by kristerkari** [comment]
Thanks for clarifying @jeddy3. I agree that that is what the configuration options should look like.
